### PR TITLE
Refactor cudf::string_view host and device code

### PR DESCRIPTION
**Pull request by davidwendt**
While working on improving the sort performance for strings columns in #7075, we tried a vector-load approach in the `string_view::compare()` function. This approached used some CUDA math intrinsic functions like `__funnelshift_r()` and `__byte_perm()`. Unfortunately, adding these to the `string_view` source would cause compile errors for some .cpp files. This is because the `string_view.cuh` was being included by some .cpp file even though these only used the appropriate `__host__ __device__` functions.

This PR breaks up the host/device from the device-only functions so the .cpp files can include `string_view.cuh` without processing the device-only definitions. The host/device functions are now defined in the `string_view.cuh` directly and the device-only source is isolated in the `string_view.inl`. The include of the `string_view.inl` is then wrapped if a `#if CUDA_ARCH` so it will not be processed by a .cpp file compilation.

Also, I attempted to minimize includes of `string_view.cuh` by removing it from `traits.hpp` and replacing it with a forward reference. This found a few files that were not including `string_view.cuh` directly as they should've. This also exposed `cpp/tests/utilities/scalar_utilities.cu` which appears to be unused and thus removed along with its header.

No functionality has changed. Build times may be slightly faster since `string_view.cuh` is included in less source files and .cpp files no longer the `string_view.inl`. This means changing this file was also have a slightly less impact on rebuilding libcudf.

### CODE REVIEW

**Review by kkraus14 — APPROVED** [review]
cmake / conda lgtm

File: cpp/include/cudf/utilities/traits.hpp [hunk 1]
```diff
@@ -505,6 +504,8 @@ constexpr inline bool is_fixed_width(data_type type)
   return cudf::type_dispatcher(type, is_fixed_width_impl{});
 }
 
+class string_view;
+
```
Comment (karthikeyann) [inline cpp/include/cudf/utilities/traits.hpp:508]: cudf/types.hpp already has forward declaration of cudf::string_view. This forward declaration may not be needed.

File: cpp/src/strings/regex/regexec.cu [hunk 2]
```diff
@@ -22,6 +22,8 @@
 #include <rmm/cuda_stream_view.hpp>
 #include <rmm/device_buffer.hpp>
 
+#include <algorithm>
+
```
Comment (karthikeyann) [inline cpp/src/strings/regex/regexec.cu:26]: copyright year update in this file.

File: cpp/tests/transform/mask_to_bools_test.cpp [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020, NVIDIA CORPORATION.
+ * Copyright (c) 2020-2021, NVIDIA CORPORATION.
```
Comment (karthikeyann) [inline cpp/tests/transform/mask_to_bools_test.cpp:2]: Do you have script or hack to update Copyright year in all files?

**Review by karthikeyann — CHANGES_REQUESTED** [review]
Neat Refactoring! LGTM. 
Minor corrections.

File: cpp/include/cudf/utilities/traits.hpp (hunk 1 above)
Comment (davidwendt) [reply cpp/include/cudf/utilities/traits.hpp:508]: Maybe, but I feel this makes the header file more complete/clear on its own.

File: cpp/tests/transform/mask_to_bools_test.cpp (hunk 3 above)
Comment (davidwendt) [reply cpp/tests/transform/mask_to_bools_test.cpp:2]: No.

**Review by ajschmidt8 — APPROVED** [review]

File: cpp/include/cudf/strings/string_view.cuh [hunk 4]
```diff
@@ -37,47 +57,28 @@ using char_utf8 = uint32_t;  ///< UTF-8 characters are 1-4 bytes
...
-  string_view(string_view&&)      = default;
-  ~string_view()                  = default;
-  string_view& operator=(const string_view&) = default;
-  string_view& operator=(string_view&&) = default;
-
   /**
    * @brief Return the number of bytes in this string
    */
-  __host__ __device__ size_type size_bytes() const;
+  __host__ __device__ inline size_type size_bytes() const { return _bytes; }
```
Comment (jrhemstad) [inline cpp/include/cudf/strings/string_view.cuh:68]: There's no need to put `inline` on member functions defined inside the class.

File: cpp/include/cudf/strings/string_view.cuh [hunk 5]
```diff
@@ -37,47 +57,28 @@ using char_utf8 = uint32_t;  ///< UTF-8 characters are 1-4 bytes
...
    * @brief Return a pointer to the internal device array
    */
-  __host__ __device__ const char* data() const;
+  __host__ __device__ inline const char* data() const { return _data; }
 
   /**
    * @brief Return true if string has no characters
    */
-  __host__ __device__ bool empty() const;
+  __host__ __device__ inline bool empty() const { return size_bytes() == 0; }
```
Comment (jrhemstad) [inline cpp/include/cudf/strings/string_view.cuh:81]: These should use `CUDA_HOST_DEVICE_CALLABLE` so the host/device decorators are removed in `.cpp` files.

File: cpp/include/cudf/strings/string_view.cuh [hunk 6]
```diff
@@ -322,31 +399,25 @@ __host__ __device__ size_type to_char_utf8(const char* str, char_utf8& character
...
+  return chr_width;
 }
 
 }  // namespace detail
 }  // namespace strings
 }  // namespace cudf
 
+// Using device specific intrinsic functions in device only code causes a
+// compile error for .cpp files.
+// Many .cpp files need only the host member functions defined above.
```
Comment (jrhemstad) [inline cpp/include/cudf/strings/string_view.cuh:418]: I'm confused, when does a .cpp file ever need to interact with a `string_view` object?

File: cpp/include/cudf/strings/string_view.cuh [hunk 7]
```diff
@@ -322,31 +399,25 @@ __host__ __device__ size_type to_char_utf8(const char* str, char_utf8& character
...
+// Therefore, the device only functions are isolated in the .inl file
+// included below.
+#if defined(__CUDA_ARCH__)
```
Comment (jrhemstad) [inline cpp/include/cudf/strings/string_view.cuh:421]: `__CUDA_ARCH__` is only valid when used inside a function. Using it at namespace scope like this is UB.

**Review by jrhemstad — CHANGES_REQUESTED** [review]

File: cpp/include/cudf/strings/string_view.cuh (hunk 6 above)
Comment (davidwendt) [reply cpp/include/cudf/strings/string_view.cuh:418]: This is because the string_view.cuh was being included by some .cpp file even though they only need to use __host__ __device__ functions and never the __device__ functions.

File: cpp/include/cudf/strings/string_view.cuh (hunk 7 above)
Comment (davidwendt) [reply cpp/include/cudf/strings/string_view.cuh:421]: I need something like this to prevent including the `.inl` file when compiled by the host compiler and to only include the `.inl` when compiled by nvcc.
Otherwise, I could add it to every `__device__` function inside the `.inl` file instead.

File: cpp/include/cudf/strings/string_view.cuh (hunk 6 above)
Comment (jrhemstad) [reply cpp/include/cudf/strings/string_view.cuh:418]: What I'm asking is when does host code ever need to use _any_ of the functions of a `string_view`?

File: cpp/include/cudf/strings/string_view.cuh (hunk 7 above)
Comment (karthikeyann) [reply cpp/include/cudf/strings/string_view.cuh:421]: > `__CUDA_ARCH__` is only valid when used inside a function. Using it at namespace scope like this is UB.

[CUDA Documentation](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#application-compatibility
) says `It is only defined for device code`. 
To be specific, it's device compilation phase. `when __CUDA_ARCH__ is defined (i.e., during device compilation phase)` 

`__CUDA_ARCH__` restrictions https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#cuda-arch-macro

nvcc can be identified using `__CUDACC__`. https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#nvcc-identification-macro 
https://stackoverflow.com/a/8809924/1550940

To summarize,
`__CUDA_ARCH__` is defined during device compilation phase (for nvcc host compilation phase, it's not defined).
`__CUDACC__` is defined only when nvcc is compiling CUDA source files. (for C/C++ files it's not defined). But it is defined for both device compilation and host compilation phase.

File: cpp/include/cudf/strings/string_view.cuh (hunk 6 above)
Comment (davidwendt) [reply cpp/include/cudf/strings/string_view.cuh:418]: One example is here creating a `string_scalar` from a `string_view`.
https://github.com/rapidsai/cudf/blob/639049808913c7d5a7170141d784da6cc45da36d/cpp/include/cudf/scalar/scalar.hpp#L430-L437
This does not require any device code. Also, getting the `string_view` from a `string_scalar`
https://github.com/rapidsai/cudf/blob/639049808913c7d5a7170141d784da6cc45da36d/cpp/include/cudf/scalar/scalar.hpp#L473-L476
And `scalar.hpp` is included in many .cpp files.
This is just an example of where a `string_view` is safely referenced/used in a .cpp file since it does not require calling any device functions.

File: cpp/include/cudf/strings/string_view.cuh (hunk 7 above)
Comment (davidwendt) [reply cpp/include/cudf/strings/string_view.cuh:421]: > __CUDACC__ is defined only when nvcc is compiling CUDA source files
This is exactly what I need.

File: cpp/include/cudf/strings/string_view.cuh (hunk 6 above)
Comment (jrhemstad) [reply cpp/include/cudf/strings/string_view.cuh:418]: Great, that makes more sense. Thanks for the examples. Those seem valid.

**Comment by davidwendt** [comment]
I decided to take a different approach than using the `__CUDA_ARCH__` define around the `string_view.inl`. 
I've renamed `string_view.cuh` to `string_view.hpp` since it contains only host or host/device code definitions.
I've renamed `string_view.inl` to `string_view.cuh` since it only contains device code.
The `string_view.cuh` includes `string_view.hpp` so if a `.cu` file needs to call any device functions, it only needs to include `string_view.cuh`. All others files can just contain `string_view.hpp`.
This removes the awkward include of a `.cuh` file in a non `.cu` file for example.
It also means we do not need the `__CUDA_ARCH__` if-define logic anymore.
I feel this is much cleaner and easier to understand.

**Review by karthikeyann — APPROVED** [review]

**Comment by harrism** [comment]
> No function as changed.

Is this a typo? Having trouble understanding what it means. :)

**Comment by davidwendt** [comment]
> > No function as changed.
> 
> Is this a typo? Having trouble understanding what it means. :)

Sorry. I meant "No function _has_ changed." I fixed the description.

**Review by jrhemstad — APPROVED** [review]
Nice solution.

**Comment by davidwendt** [comment]
@gpucibot merge
